internal/dataplane/kongstate/plugin.go

```diff
@@ -4,18 +4,122 @@ import (
 	"encoding/json"
 	"errors"
 	"fmt"
+	"strings"
 
 	jsonpatch "github.com/evanphx/json-patch/v5"
 	"github.com/kong/go-kong/kong"
+	"github.com/samber/lo"
 	corev1 "k8s.io/api/core/v1"
 	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/yaml"
 
 	"github.com/kong/kubernetes-ingress-controller/v3/internal/store"
 	"github.com/kong/kubernetes-ingress-controller/v3/internal/util"
 	kongv1 "github.com/kong/kubernetes-ingress-controller/v3/pkg/apis/configuration/v1"
 )
 
+// Plugin represents a plugin Object in Kong.
+type Plugin struct {
+	kong.Plugin
+	K8sParent           client.Object
+	SensitiveFieldsMeta PluginSensitiveFieldsMetadata
+}
+
+func (p Plugin) DeepCopy() Plugin {
+	return Plugin{
+		Plugin:              *p.Plugin.DeepCopy(),
+		K8sParent:           p.K8sParent,
+		SensitiveFieldsMeta: p.SensitiveFieldsMeta,
+	}
+}
+
+func (p Plugin) SanitizedCopy() Plugin {
+	// We do not want to return an error if any of below fails - the best we can do
+	// is to return a plugin with wholly redacted config.
+	// Let's have a closure returning a plugin with wholly redacted config prepared.
+	whollySanitized := func() Plugin {
+		p := p.DeepCopy()
+		p.Config = sanitizeWholePluginConfig(p.Config)
+		return p
+	}
+
+	// If the whole config is sensitive, we need to redact the entire config.
+	if p.SensitiveFieldsMeta.WholeConfigIsSensitive {
+		return whollySanitized()
+	}
+
+	// If there are JSON paths, we need to redact them.
+	if len(p.SensitiveFieldsMeta.JSONPaths) > 0 {
+		var patchOperations []string
+		for _, path := range p.SensitiveFieldsMeta.JSONPaths {
+			// If the path is empty, we need to sanitize the whole config.
+			// An empty path means that the patch is on the root of the config.
+			if path == "" {
+				return whollySanitized()
+			}
+
+			patchOperations = append(patchOperations, fmt.Sprintf(
+				`{"op":"replace","path":"%s","value":"%s"}`,
+				path,
+				*redactedString,
+			))
+		}
+
+		// Decode the patch and apply it to the config.
+		// We need to marshal the config to JSON and then unmarshal it back to Configuration
+		// because the patch library works with bytes.
+		patch, err := jsonpatch.DecodePatch([]byte(fmt.Sprintf("[%s]", strings.Join(patchOperations, ","))))
+		if err != nil {
+			return whollySanitized()
```

Comment:
I wonder whether we should log something in such a case, ofc with DEBUG or TRACE level. But it's not something that I have a strong opinion about. For your own judge 😉

Reply:
Yeah, I was considering that but ended up leaving it as is because I concluded that those error cases are very unlikely to happen and even if they happen, it won't affect a user much - it will just make the whole config to be sanitized which is still a valid outcome.